Entypo+ and associated defs, many new docs, colorpicker, updated examples

svenevs opened the pull request:
Definitely squash this, there was a lot of intermediate commits for the entypo stuff that have now been reverted or destroyed.

~~**NOTE**: I'm going through and doing a PR review right now.  Don't review until this is gone since this is a huge update and most of it is just noise from doc comments.~~

1. It would be kind of tedious to separate this out into multiple PRs, but it can be done if you prefer.  Things like docs links got added in a few different locations relating to entypo.

2. See annotations in PR review.  Basically, this all boils down to the following:

    - Entypo+ updates
        - Some constants have changed, those are all fixed.  Entypo+ file documentation enumerates everything, as well as there are CPP and python examples.
        - Python binding code changes slightly to allow auto-generation of Entypo+ constants without having to incorporate GLFW stuff in the gen code.
        - **Important**: icon sizes have changed.  See [relevant change to common.h](https://github.com/wjakob/nanogui/pull/260#discussion_r139310396).  Here is a full index of where that applies:
            - [first scale change](https://github.com/wjakob/nanogui/pull/260#discussion_r139310615)
            - [second scale change](https://github.com/wjakob/nanogui/pull/260#discussion_r139310622)
            - [third scale change](https://github.com/wjakob/nanogui/pull/260#discussion_r139310634)
            - [fourth scale change](https://github.com/wjakob/nanogui/pull/260#discussion_r139310724)
            - [fifth scale change](https://github.com/wjakob/nanogui/pull/260#discussion_r139310729)
            - [sixth scale change](https://github.com/wjakob/nanogui/pull/260#discussion_r139310735)
            - [seventh scale change](https://github.com/wjakob/nanogui/pull/260#discussion_r139310748)
            - Basically, the ones using `ih` should be good.  But a couple were not scaled the same as previous code.
    - ColorPicker has been modified
        - I believe the new functionality makes sense.  These changes came about from writing the docs and not understanding how it worked.  The callback gets executed too frequently.  Since it's used in the form helper, I decided it needed to change ;)
    - Sad face.  Docs required getting rid of `using` for eigen vector types :/

3. I'm going to split off onto another branch now for the remaining docs updates.  I made a serializer example, but that and the changes made for the docs will be in their own self-contained PR because they aren't exactly trivial changes.

The rest is just docs updates really.  Highlight of fake docs build [entypo.h](http://fake-nanogui.readthedocs.io/en/latest/api/file_nanogui_entypo.h.html#file-nanogui-entypo-h) :)

P.S. In future merge, you may need to **wipe** the environment for RTD to build it right because of the change in `requirements.txt` and more specifically if `exhale.pyc` is still floating around it will complain about no `setup(app)` "are you sure it is a sphinx extension" because it wasn't.

svenevs commented on CMakeLists.txt:
```diff
@@ -433,7 +435,8 @@ if (NANOGUI_BUILD_PYTHON)
   add_library(nanogui-python-obj OBJECT
     python/main.cpp
-    python/constants.cpp
+    python/constants_glfw.cpp
```
`python/constants.cpp` got split into two files.

1. `constants_glfw.cpp` has same as before, just in a different file.
2. `constants_entypo.cpp` is generated (which is why they got split)

wjakob replied:
Good call

svenevs commented on docs/conf.py:
```diff
@@ -32,14 +32,85 @@
+    ############################################################################
+    "createTreeView":        True,
+    "exhaleExecutesDoxygen": True,
+    "exhaleDoxygenStdin": textwrap.dedent('''
```
For the most part, all the other Doxygen stuff from the original `conf.py` are in the Exhale defaults.  We only need `INPUT` and `PREDEFINED` for the namespaces.  I still have no idea why Doxygen can't handle those.

svenevs commented on docs/conf.py:
```diff
@@ -32,14 +32,85 @@
+    # Fix broken Sphinx RTD Theme 'Edit on GitHub' links
+    # Search for 'Edit on GitHub' on the FAQ:
+    #     http://exhale.readthedocs.io/en/latest/faq.html
+    "pageLevelConfigMeta": ":github_url: https://github.com/wjakob/nanogui",
```
This is the (temporary, but not going to be fixed any time soon) hack to get the generated pages to link back to a real URL.

svenevs commented on docs/conf.py:
```diff
@@ -303,102 +366,7 @@
-    from exhale import generate
-    generate(exhaleArgs)
+    # this looks in html_static_path (_static directory as set above)
+    app.add_stylesheet("entypo.css")
```
Note: `fontcustom` generated `entypo.css` will **NOT** conflict with sphinx theme because all CSS targets are `entypo-icon-X` rather than just `icon-X`.  AKA even though this is linked to on every page, it doesn't cause issues.

wjakob replied:
Having the icons in the HTML docs is super-nifty. I always dreaded finding the right symbols with boring recompilation cycles.

svenevs replied:
LOL yeah it was a little tedious to get started, but the CSS generated can be parsed with regular expressions so once that was done it was good

svenevs commented on docs/examples.rst:
```diff
@@ -58,6 +60,24 @@ arbitrary sequence of OpenGL commands into a NanoGUI widget.
 
+.. _nanogui_example_icons:
+
+Example Icons
```
These were added for testing, but I don't see the harm in having them.  The icons will show up slightly differently than the webpage because of the default dark background of NanoGUI.  Can remove if undesired.

wjakob replied:
Neat. Would you mind calling the executable ``example_icons``?

svenevs replied:
Done

svenevs commented on docs/index.rst:
```diff
@@ -64,7 +64,7 @@ Contents
    compilation
    examples
-   generated_api/library_root
+   api/library_root
```
This change means that people who have setup non intersphinx links will now direct to invalid URLs.  But the new url makes more sense, as well as `exhale_` is removed from the beginning of every filename as it now requires a subdirectory in the sphinx source dir.  `exhale_` prefix was because it wasn't a strict requirement before.

svenevs commented on docs/requirements.txt:
```diff
@@ -1,2 +1 @@
-breathe
-# git+git://github.com/michaeljones/breathe@v4.3.1#egg=breathe
+exhale
```
To build locally, you'll need to `pip3 install exhale`.  NOTE: py3 is noticeably faster than py2 ;)

wjakob replied:
Nice to hear -- many times, it is unfortunately the other way ;).

svenevs commented on include/nanogui/colorpicker.h:
```diff
@@ -21,24 +22,84 @@ NAMESPACE_BEGIN(nanogui)
     void setColor(const Color& color);
+
+    /// The current caption of the \ref nanogui::ColorPicker::mPickButton.
+    const std::string &pickButtonCaption() { return mPickButton->caption(); }
```
Added pass-through methods for setting button captions if desired.

svenevs commented on include/nanogui/colorpicker.h:
```diff
@@ -21,24 +22,84 @@ NAMESPACE_BEGIN(nanogui)
+     * caption of this Button is ``"Reset"``.  You can change it using
+     * \ref nanogui::ColorPicker::setResetButtonCaption if you need.
+     */
+    Button *mResetButton;
```
This is new.  The changed functionality of ColorPicker necessitates this.  See comments in cpp file.

svenevs commented on include/nanogui/common.h:
```diff
@@ -109,39 +111,47 @@ struct GLFWcursor;
-using Eigen::VectorXf;
-using Eigen::MatrixXf;
+/// Type alias to allow ``Eigen::Vector2f`` to be used as ``nanogui::Vector2f``.
+typedef Eigen::Vector2f Vector2f;
```
Breathe chokes on raw `using` in a namespace (it does seem to do ok with them in a class).  Without explicit typedefs here, the fact that `nanogui::Vector3f` etc are available never show up in the docs.

This can easily be undone, at the expense of no explicit documentation of the vector types.

wjakob replied:
Typedefs are so C++03 :(. Can we at least do ``using VectorXf = Eigen::VectorXf``, etc?

svenevs replied:
I'll look into `using` tomorrow, I think I tried that and it broke but don't remember.  I know `using x = thing` is ok in classes though.

(sorry if you get a lot of emails from edits, the github review thing is nice but i overcommented the first one.......)

svenevs commented on include/nanogui/common.h:
```diff
@@ -494,7 +504,13 @@ extern NANOGUI_EXPORT std::vector<std::pair<int, std::string>>
 
+namespace detail {
+    /// Used for scaling icons (primarily Entypo) in widget draw calls.
+    static constexpr float __icon_scale = 0.77f;
```
The new Entypo+ font is presumably bigger.  See annotations in various CPP files for how this is used.

wjakob replied:
Sounds fair. I don't think it needs to have such a cryptic name (``__icon_scale``) and be ``constexpr``. The ``Theme`` class would seem like a more natural home to me.

wjakob replied:
I just realized that ``entypo`` has a CC-BY-SA license, which is a bit more restrictive than the BSD license that is used for the rest of NanoGUI. Unlikely to be an issue to anyone, but it may be worth adding a note somewhere in the docs.

svenevs replied:
Should it go on the README?

svenevs replied:
As for cryptic name, I'm adding it to the theme.  The `__` came from `__nanogui_get_image` above it...but I see now that's because of the `define` above xD

svenevs commented on include/nanogui/glcanvas.h:
```diff
@@ -21,47 +25,67 @@ NAMESPACE_BEGIN(nanogui)
  *
- * Usage: override `drawGL` in subclasses to provide custom drawing code.
+ * \rst
+ * **Usage**
```
For the most part, all of @fairlight1337 's comments stayed the same.  I just linked back to the example, and put some `.`'s at the end of doc comments because I've been obsessively doing that to everything.  AKA these doc changes are just for consistency, but I also added that he contributed the widget :)

wjakob replied:
Sounds good :)

svenevs commented on include/nanogui/opengl.h:
```diff
@@ -15,6 +15,10 @@
 
+// The Entypo PLUS and MINUS icons are 43 and 45 respectively, which needs to
+// be checked manually.
+#include <nanogui/entypo.h>
```
Ooops.  Need to remove this, will happen in next commit on this branch.  This was the weird font encoding issue, was comparing to those two icon values explicitly in the helper functions.

svenevs commented on include/nanogui/textbox.h:
```diff
@@ -66,8 +66,10 @@ class NANOGUI_EXPORT TextBox : public Widget {
     virtual void setTheme(Theme *theme) override;
 
-    /// Set the change callback
+    /// The callback to execute when the value of this TextBox has changed.
```
The rest of `textbox.h` is still a TODO.  I only updated things to make docs link back and forth.  Same happened in `widget.h`.

svenevs commented on python/constants.cpp:
```diff
@@ -1,107 +0,0 @@
-#ifdef NANOGUI_PYTHON
```
As discussed above, this file is now gone in favor of two (one for GLFW, one for Entypo)

svenevs commented on python/eigen.cpp:
```diff
@@ -16,6 +16,8 @@ void register_eigen(py::module &m) {
                       [](Color &c, float v) { c.g() = v; }, D(Color, g))
         .def_property("b", [](const Color &c) { return c.b(); },
                       [](Color &c, float v) { c.b() = v; }, D(Color, b))
+        .def_property("w", [](const Color &c) { return c.w(); },
```
This one needs to be bound with an explicit docstring because `make mkdoc` will not get the inherited Eigen docs (which is where `Vector4f::w()` comes from).  This is used in the new examples.

svenevs commented on python/example1.py:
```diff
@@ -278,32 +278,19 @@ def cb(value):
-        colorBtn.setFixedSize((100, 25))
-        c = colorwheel.color()
-        colorBtn.setBackgroundColor(c)
+        Label(window, "Color picker :", "sans-bold");
```
Example 1 in python; Entypo icon fixes and using ColorPicker directly.

svenevs commented on python/example2.py:
```diff
@@ -55,7 +55,16 @@ def getter():
+                                                   color.w)
+    )
+
+gui.addColorVariable("Color", *make_accessors("colvar")).setCallback(cp_cb)
```
FormHelper uses ColorPicker, add callback to demonstrate its new functionality.

svenevs commented on python/exampleIcons.py:
```diff
@@ -0,0 +1,62 @@
+    # NOTE: don't __dict__ crawl in real code!
+    # this is just because it's more convenient to do this for enumerating all
+    # of the icons -- see cpp example for alternative...
+    for key in entypo.__dict__.keys():
```
I'd rather not do what was done in C++, but since this is an example, maybe this is bad?

wjakob replied:
It seems Pythonic enough to me, definitely preferable to the C++ version.

svenevs commented on python/main.cpp:
```diff
@@ -37,7 +37,8 @@ namespace {
 #endif
 
-extern void register_constants(py::module &m);
+extern void register_constants_glfw(py::module &m);
```
The corresponding updates to the python bindings.

svenevs commented on src/button.cpp:
```diff
@@ -31,7 +31,7 @@ Vector2i Button::preferredSize(NVGcontext *ctx) const {
     if (mIcon) {
         if (nvgIsFontIcon(mIcon)) {
-            ih *= 1.5f;
+            ih *= detail::__icon_scale;
```
First icon scaling change.

svenevs commented on src/button.cpp:
```diff
@@ -169,7 +169,7 @@ void Button::draw(NVGcontext *ctx) {
         float iw, ih = fontSize;
         if (nvgIsFontIcon(mIcon)) {
-            ih *= 1.5f;
+            ih *= detail::__icon_scale;
```
Second icon scaling change.

svenevs commented on src/checkbox.cpp:
```diff
@@ -78,7 +78,7 @@ void CheckBox::draw(NVGcontext *ctx) {
 
     if (mChecked) {
-        nvgFontSize(ctx, 1.8 * mSize.y());
+        nvgFontSize(ctx, detail::__icon_scale * mSize.y());
```
Third icon scaling change.  I got lazy and used one globally, this one may need to change.

wjakob replied:
I liked the bigger check mark better. Sounds like you'll need an extra factor of ``1.2f``.

svenevs commented on src/colorpicker.cpp:
```diff
@@ -22,26 +22,53 @@ ColorPicker::ColorPicker(Widget *parent, const Color& color) : PopupButton(paren
-        mCallback(backgroundColor());
+        if (this->mPickButton->pushed()) {
+            setColor(backgroundColor());
+            mCallback(backgroundColor());
```
Callback was getting executed too frequently, e.g. when a user changes the color wheel value but does NOT select "Pick".  Remaining changes in this file relate to this, as well as the reset button.

wjakob replied:
One option would be to split that into two callbacks. For instance, the slider widget has a fast callback and a slow one, which may be useful for quick interactive visualizations.

svenevs replied:
I'm not sure I understand.  If you change the current `master` of NanoGUI example1 to use the `ColorPicker`, it might be clearer why this change is here.  I rewrote `ColorPicker` to match the existing example1 functionality, specifically that the button color does not change until it's been selected.  The reset button was added (can be removed) to allow the color wheel to be reset to the actively selected color.

How does a different callback come in?

wjakob replied:
Maybe it's me who is misunderstanding. I thought that it would be nice to have interactive feedback when e.g. changing the color of a mesh using the color picker. That requires a fast callback (i.e. "user is currently tweaking the color") and a slow one ("Pick/Reset button was finally pressed"). Potentially it will make things too complicated though.

svenevs commented on src/example1.cpp:
```diff
@@ -458,31 +459,15 @@ class ExampleApplication : public nanogui::Screen {
-                popupBtn->setBackgroundColor(colorBtn->backgroundColor());
-                popupBtn->setPushed(false);
-            }
+        new Label(window, "Color picker :", "sans-bold");
```
Corresponding example1 updates in CPP.

svenevs commented on src/example2.cpp:
```diff
@@ -63,7 +63,14 @@ int main(int /* argc */, char ** /* argv */) {
            ->setItems({"Item 1", "Item 2", "Item 3"});
-        gui->addVariable("Color", colval);
+        gui->addVariable("Color", colval)
+           ->setCallback([](const Color &c) {
```
Example 2 updates in CPP.

svenevs commented on src/example3.cpp:
```diff
@@ -113,7 +113,14 @@ int main(int /* argc */, char ** /* argv */) {
     gui->addGroup("Complex types");
     gui->addVariable("Enumeration", enumval, enabled)->setItems({ "Item 1", "Item 2", "Item 3" });
-    gui->addVariable("Color", colval);
+    gui->addVariable("Color", colval)
```
Example 3 updates in CPP.  Does not apply to python example 3.

svenevs commented on src/popupbutton.cpp:
```diff
@@ -45,7 +45,7 @@ void PopupButton::draw(NVGcontext* ctx) {
             mTextColor.w() == 0 ? mTheme->mTextColor : mTextColor;
 
-        nvgFontSize(ctx, (mFontSize < 0 ? mTheme->mButtonFontSize : mFontSize) * 1.5f);
+        nvgFontSize(ctx, (mFontSize < 0 ? mTheme->mButtonFontSize : mFontSize) * detail::__icon_scale);
```
Fourth icon scale change.  This one used the original `ih *= 0.5` values, but it may need a different scale.

svenevs commented on src/tabheader.cpp:
```diff
@@ -407,10 +407,10 @@ void TabHeader::drawControls(NVGcontext* ctx) {
     int fontSize = mFontSize == -1 ? mTheme->mButtonFontSize : mFontSize;
     float ih = fontSize;
-    ih *= 1.5f;
+    ih *= detail::__icon_scale;
```
Fifth icon scale.  Should be good.

svenevs commented on src/tabheader.cpp:
```diff
@@ -429,10 +429,10 @@ void TabHeader::drawControls(NVGcontext* ctx) {
     fontSize = mFontSize == -1 ? mTheme->mButtonFontSize : mFontSize;
     ih = fontSize;
-    ih *= 1.5f;
+    ih *= detail::__icon_scale;
```
Sixth icon scale.

svenevs commented on src/textbox.cpp:
```diff
@@ -151,7 +151,7 @@ void TextBox::draw(NVGcontext* ctx) {
 
         nvgFontFace(ctx, "icons");
-        nvgFontSize(ctx, ((mFontSize < 0) ? mTheme->mButtonFontSize : mFontSize) * 1.2f);
+        nvgFontSize(ctx, ((mFontSize < 0) ? mTheme->mButtonFontSize : mFontSize) * detail::__icon_scale);
```
Seventh icon scale.  This one is not the "default" `1.5`, and may also need a different value?

wjakob replied:
Yep, the little scroll handles got bigger. It looked nicer before.

wjakob commented:
This looks great, I added a few minor comments.

wjakob commented:
One more thing: somehow the top icons of the tab widget look messed up for me (try ``example1``). Does it happen on your machine as well?

svenevs commented:
Ok icon scaling attempt redone.  I'm a little concerned about positioning, but I think the easiest solution is to avoid `SMALL` icons.  The alternative solution was too confusing for me, but it would be to somehow account for `Widget::icon_scale()` in the width positioning.  I got really turned around trying that though...

If you use this branch and set change `popupbutton.cpp`'s constructor at the top to use `ENTYPO_ICON_CHEVRON_SMALL_RIGHT` and comment out `mIconExtraScale`, Example1 specifically on the Combo Box is wrong.  I added a left popup to make sure nothing went wrong there, and I think the underlying symptom is that `Combo box item 1` is starting to get too big?  Basically, I'm not sure if this is a hidden bug, or if current scaling approach is good enough.

----------

I'll look into `using` tomorrow, I think I tried that and it broke but don't remember.  I know `using x = thing` is ok in classes though.

------------

> One more thing: somehow the top icons of the tab widget look messed up for me (try example1). Does it happen on your machine as well?

Do you mean the left and right arrows or the `+` for the dynamic add?

wjakob commented:
Sure, feel free to delete the other comments. Yes, I'm talking about the left and right buttons to scroll through a longer list of tabs.

svenevs commented:
> Yes, I'm talking about the left and right buttons to scroll through a longer list of tabs.

I'm hoping you were only talking about **python** example 1.  The dynamic stuff wasn't in that example (see latest commit), so the icons were not displayed because there was nowhere else to scroll in the tab widget.  If you're talking about CPP this is a problem.

ColorPicker updated to have `callback` and `finalCallback` like the slider.  New window added to Example 1 to show that.

-------------------

Assuming those changes are good, the last items before this is ready are

1. Explicitly notify users about the CC-BY-SA license for Entypo+.  Proposed location is on the main README, and included in `index.rst` from the README.
2. `typedef` vs `using`.  ColorPicker fixes took longer than expected, should be able to add these two later tonight (tomorrow for you, I'm in UTC-8 right now).

svenevs commented:
> 1. Explicitly notify users about the CC-BY-SA license for Entypo+. Proposed location is on the main README, and included in index.rst from the README.

See [updated `README.rst`](https://github.com/svenevs/nanogui/tree/docs-redux).  Nice thing about switching to rst is `.. contents::` works on GitHub, and now there's less duplicated text (still some though) :smile:

Please review latest commit, I moved everything to Theme because after the research I did on CC BY-SA it wasn't clear exactly how attribution should happen.  @nhojb this may or may not apply to you with ColorFinale.  In his [FAQ](http://entypo.com/faq.php) he says you could even put a comment in the source code, but that really only adheres to SA for webpages.  Not my domain...

Anyway, with all used icons being in Theme, if it is a problem then people can include a different icon font.  I added how to do that, as well as some other cmake hacks in the docs (`compilation.rst`).

> 2. typedef vs using. ColorPicker fixes took longer than expected, should be able to add these two later tonight (tomorrow for you, I'm in UTC-8 right now).

I went with `using Vector3f = Eigen::Vector3f`, but it does have a consequence.  The rendered docs give something like `using Vector3f = template Eigen::Vector3f`.  [It's been around for a while](https://github.com/michaeljones/breathe/issues/316), but `breathe` is sort of going stale.  The old-school `typedefs` work fine, and some `using` do.  I may investigate that in January.

Long story short, the current approach is good enough.  It adds build warnings and there's a weird signature, but the important part is that the docs are linkable so that when something returns a `Vector2i`, users can click on the return type and it will take them to the typedef page.  The docstrings are clear enough that it shouldn't be an issue.

wjakob commented:
Great -- that sounds good, and no worries about the ``breathe`` issue, it seems fairly minor. How would you like this merged? Do you want to squash it down into a smaller number of commits?

svenevs commented:
Hmm. Yeah squashing into a single commit is probably a bad idea.  I'm attaching the full [260_patch.txt](https://github.com/wjakob/nanogui/files/1319732/260_patch.txt) here.
I'll split this out into three commits: Entypo, ColorPicker, and docs.  Because I don't remember if GitHub gets mad about rebasing PRs manually, I'll create a backup branch too...

svenevs commented:
Wow that was actually kind of hard.  I gave up on trying to `rebase` and just did some manual copy ops with some `git checkout HEAD -- ....` LOL.  So right now my fork has a `docs-redux-backup` branch, here's the final diff between that and the three commits here now:

```diff
diff --git a/python/example1.py b/python/example1.py
index 2bea681..237465f 100644
--- a/python/example1.py
+++ b/python/example1.py
@@ -16,7 +16,7 @@ import gc
 
 from nanogui import Color, ColorPicker, Screen, Window, GroupLayout, BoxLayout, \
                     ToolButton, Label, Button, Widget, \
-                    PopupButton, CheckBox, MessageDialog, VScrollPanel, \
+                    Popup, PopupButton, CheckBox, MessageDialog, VScrollPanel, \
                     ImagePanel, ImageView, ComboBox, ProgressBar, Slider, \
                     TextBox, ColorWheel, Graph, GridLayout, \
                     Alignment, Orientation, TabWidget, IntBox, GLShader
@@ -75,10 +75,17 @@ class TestApp(Screen):
         popup.setLayout(GroupLayout())
         Label(popup, "Arbitrary widgets can be placed here")
         CheckBox(popup, "A check box")
+        # popup right
         popupBtn = PopupButton(popup, "Recursive popup", entypo.ICON_FLASH)
-        popup = popupBtn.popup()
-        popup.setLayout(GroupLayout())
-        CheckBox(popup, "Another check box")
+        popupRight = popupBtn.popup()
+        popupRight.setLayout(GroupLayout())
+        CheckBox(popupRight, "Another check box")
+        # popup left
+        popupBtn = PopupButton(popup, "Recursive popup", entypo.ICON_FLASH)
+        popupBtn.setSide(Popup.Side.Left)
+        popupLeft = popupBtn.popup()
+        popupLeft.setLayout(GroupLayout())
+        CheckBox(popupLeft, "Another check box");
 
         window = Window(self, "Basic widgets")
         window.setPosition((200, 15))
diff --git a/src/example1.cpp b/src/example1.cpp
index d06f022..545c5c0 100644
--- a/src/example1.cpp
+++ b/src/example1.cpp
@@ -198,7 +198,6 @@ public:
         // popup left
         popupBtn = new PopupButton(popup, "Recursive popup", ENTYPO_ICON_FLASH);
         popupBtn->setSide(Popup::Side::Left);
-        popupBtn->setChevronIcon(ENTYPO_ICON_CHEVRON_LEFT);
         Popup *popupLeft = popupBtn->popup();
         popupLeft->setLayout(new GroupLayout());
         new CheckBox(popupLeft, "Another check box");
```

Thus I believe this was a success :laugh:

wjakob commented:
Merged (& wiped RTD) -- thank you very much for the epic PR!

wjakob commented:
Hmm -- I just noticed that this PR still breaks the tab widget. See 

https://www.dropbox.com/s/9txq8y980tomvlz/Screen%20Shot%202017-09-22%20at%2011.23.15.png?dl=0

It used to have arrows on both sides -- would you mind fixing this?

svenevs commented:
Uh oh. Maybe this should be reverted. What OS are you using, Ubuntu?

wjakob commented:
It's on Mac OS X.

svenevs commented:
Ugh. Yeah I'm gonna have to dig a little. Main testing on my part was osx Sierra and fedora

svenevs commented:
See #263 I'm really not sure what's going on here...

svenevs commented:
@wjakob Eston indicated to me that the search index on RTD is somehow out of date.  E.g. if you search for `GLShader` it'll somehow link to the old one.  The only thing I can think of is re-wipe and re-launch a build.  The fake nanogui site search index is up to date, so this may just be a fluke.
